### PR TITLE
Solving the tasks for Module #3

### PR DESCRIPTION
**Pull request by NatalliaBileichyk**
1. Solve the tasks for Module #3 of the js-webinar project.
2. Solve the tasks for Plus >OOP module

### CODE REVIEW

File: module-3/pop/HomePage.js [hunk 1]
```diff
@@ -7,16 +7,25 @@ const Layout = require('./Layout');
...
-        this.menuItems = new Elements("Menu Items", by.css('.top-navigation__item-link'));
+class HomePage extends Layout {
+    constructor(name, url, locator) {
+        super(name, url, locator);
     }
 
-    get logo() {
-        return this.get('Logo');
+    getLogo () {
+        const newElementLogo = new Element('Logo', {css:'.header__logo'});
```
Comment (rakur) [inline module-3/pop/HomePage.js]: These elements should be children of the page, and should not be created when calling a get function

File: module-3/test/pop/HomePage.spec.js [hunk 2]
```diff
@@ -1,3 +1,64 @@
...
+        delete global.element;
+        delete global.browser;
+    });
+    
+    it('should be defined', () => {
+        expect(HomePage).to.be.instanceOf(Function);
+    });
+
+    it('should have the page name', () => {
+        const page = new HomePage('EPAM HomePage', 'http://epam.com', {css: 'body'});
```
Comment (rakur) [inline module-3/test/pop/HomePage.spec.js]: Can we have just one homepage?

File: plus/oop/circle.js [hunk 3]
```diff
@@ -5,4 +5,31 @@
  * @property {number} r the radius of the circle
  * @function area return the area of the circle
  * @function perimeter return the perimeter of the circle
- */
\ No newline at end of file
+ */
+
+class Circle {
+    constructor (number, r){
```
Comment (rakur) [inline plus/oop/circle.js]: Do you need this many parameters and properties?

File: plus/oop/circle.js [hunk 4]
```diff
@@ -5,4 +5,31 @@
...
+        this.r = number;
+
+        //should handle invalid radius
+        if (this.r <= 0) {
+            throw new Error ("Invalid radius")
+        }
+    }
+        
+    //should calculate perimeter
+    getPerimeter (number){
```
Comment (rakur) [inline plus/oop/circle.js]: Do you need a parameter here?

File: plus/oop/circle.js [hunk 5]
```diff
@@ -5,4 +5,31 @@
...
+        return this.number*2*Math.PI;
+    }
+        
+    // should calculate ares
+    getArea(number){
```
Comment (rakur) [inline plus/oop/circle.js]: Do you need a parameter here?

File: plus/oop/delete.js [hunk 6]
```diff
@@ -6,4 +6,17 @@
...
+
+function del (o, key) {
+        
+    if (typeof o !== "object") {
+        throw new Error ("No object is passed")
+    }
+
+    const newObject = Object.assign({}, o);
+    //return Reflect.deleteProperty(newObject,key);
+    return delete(newObject,key);
```
Comment (rakur) [inline plus/oop/delete.js]: Is there such a function?

File: plus/oop/properties.js [hunk 7]
```diff
@@ -4,4 +4,19 @@
...
- */
\ No newline at end of file
+ */
+
+function properties (o) {
+    // should handle if parameter is not object
+
+    // if (typeof o !== "object") {return 0} - doesn't work. WHY?
+    // if (!o) {return 0}
+    if (!Object.keys(o)) {
```
Comment (rakur) [inline plus/oop/properties.js]: Should return an empty array, if the parameter is not an object

File: plus/oop/circle.js (hunk 3 above)
Comment (NatalliaBileichyk) [reply plus/oop/circle.js]: Unnecessary parameter is removed

File: plus/oop/circle.js (hunk 4 above)
Comment (NatalliaBileichyk) [reply plus/oop/circle.js]: Unnecessary parameter is removed

File: plus/oop/circle.js (hunk 5 above)
Comment (NatalliaBileichyk) [reply plus/oop/circle.js]: Unnecessary parameter is removed

File: plus/oop/properties.js (hunk 7 above)
Comment (NatalliaBileichyk) [reply plus/oop/properties.js]: Return is rewritten.

File: plus/oop/delete.js (hunk 6 above)
Comment (NatalliaBileichyk) [reply plus/oop/delete.js]: I rewrite the return and check it on my examples (use without spec.js).

File: module-3/pop/HomePage.js (hunk 1 above)
Comment (NatalliaBileichyk) [reply module-3/pop/HomePage.js]: Constants have been added to the constructor

File: module-3/test/pop/HomePage.spec.js (hunk 2 above)
Comment (NatalliaBileichyk) [reply module-3/test/pop/HomePage.spec.js]: The "page" constant has been set once for all test cases.

File: plus/oop/delete.js [hunk 8]
```diff
@@ -6,4 +6,15 @@
...
\ No newline at end of file
+ */
+
+function del (o, key) {    
+    if (typeof o !== "object") {
+        throw new Error ("No object is passed")
+    }
+
+    const newObject = Object.assign({}, o);
+    return delete newObject[key]
```
Comment (rakur) [inline plus/oop/delete.js]: delete operator returns the whether the deletion was successful or not, not the object without the property

File: module-3/pop/HomePage.js [hunk 9]
```diff
@@ -7,16 +7,31 @@ const Layout = require('./Layout');
...
-        this.header.addChildren(new Element("Logo", by.css('.header__logo')));
-        this.menuItems = new Elements("Menu Items", by.css('.top-navigation__item-link'));
+class HomePage extends Layout {
+    constructor(name, url, locator) {
+        super(name, url, locator);
+        
+        this.newElementHeader = new Element('Header', {css:'header'});
+        this.newElementLogo = new Element('Logo', {css:'.header__logo'});
+        this.newElementMenuItems = new Elements("Menu Items", {css:'.top-navigation__item-link'});
+        this.newElementFooter = new Elements("Footer", {css:'.footer__holder'});
```
Comment (rakur) [inline module-3/pop/HomePage.js]: Are you sure this is a collection, and should be of class Elements?

File: module-3/pop/HomePage.js [hunk 10]
```diff
@@ -7,16 +7,31 @@ const Layout = require('./Layout');
...
     }
 
-    get logo() {
-        return this.get('Logo');
+    getLogo () {
```
Comment (rakur) [inline module-3/pop/HomePage.js]: Getters named like this should return the object itself not just a property, either rename the function, or change the return value

Comment (NatalliaBileichyk) [reply module-3/pop/HomePage.js]: Methods and parameters have been renamed.

File: module-3/pop/HomePage.js (hunk 9 above)
Comment (NatalliaBileichyk) [reply module-3/pop/HomePage.js]: As the footer contains several elements I think it should be a collection.

File: plus/oop/delete.js (hunk 8 above)
Comment (NatalliaBileichyk) [reply plus/oop/delete.js]: done

File: module-3/pop/HomePage.js [hunk 11]
```diff
@@ -7,16 +7,31 @@ const Layout = require('./Layout');
...
-        this.header = new Element('Header', by.css('header'));
-        this.header.addChildren(new Element("Logo", by.css('.header__logo')));
-        this.menuItems = new Elements("Menu Items", by.css('.top-navigation__item-link'));
+    retrieveLogo() {
+        return this.logo.name; 
     }
 
-    get logo() {
-        return this.get('Logo');
+    retrieveHeader () {
```
Comment (rakur) [inline module-3/pop/HomePage.js:25]: You have the get() function for these, please use that

**Review by rakur — APPROVED** [review]

**Pull request merged**